scylla/src/transport/load_balancing/mod.rs

```diff
@@ -92,6 +92,62 @@ fn slice_rotated_left<T>(slice: &[T], mid: usize) -> impl Iterator<Item = &T> +
     begin.iter().chain(end.iter())
 }
 
+/// This allows the optimisation in case of statement being a LWT: if so,
+/// the statements' queries are routed to the replicas only
+/// (as it speeds LWT execution up significantly).
+pub(crate) fn make_plan_considering_lwt<'a>(
```

Comment:
I think this is an overkill. You are forcing token awareness upon users which don't want it. Moreover, you aren't taking DC awareness into consideration.

The implementation in gocql just made sure that the token aware host policy always returns replicas in the specified order. Would there be some problems if we implemented it that way? The only one I can see is that a token aware policy could be composed with another policy in such a way that the parent policy breaks this ordering in some bad way.

Reply:
I believe I've followed Java driver implementation: https://github.com/scylladb/java-driver/blob/ed54dad73c96c724d8a4e28da9dc26893c031313/driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java#L135
If this implementation is bad, perhaps we should change it in Java driver as well?

Reply:
I'm not entirely sure. I read the discussion about the LWT optimization in the gocql repo and it turns out that the Java behavior might be better when using SERIAL consistency whereas the implementation in gocql should work better with LOCAL_SERIAL.

Ref: https://github.com/scylladb/gocql/issues/40#issuecomment-616740156

I'm not sure whether the issue was considered during development of the support in the Java driver, however I think the gocql behavior a bit more sense as the user retains more control over how replicas are chosen.

@avelanarius @Lorak-mmk could you comment on the current behavior of the Java driver?

Reply:
> The implementation in gocql just made sure that the token aware host policy always returns replicas in the specified order. Would there be some problems if we implemented it that way?

What you say is about covering the case of LWTs in use cases when a token aware policy is used. What if it is not used, and instead there is a round robin policy, which essentially contradicts the idea of preserving the specified order?

Reply:
I'd leave the round robin policy as it is. It isn't very useful by itself and is usually used as a child policy (e.g. along with token aware policy).